nndownload/downloader.py

```diff
@@ -0,0 +1,53 @@
+import re
+from concurrent.futures import ThreadPoolExecutor
+
+from Crypto.Cipher import AES
+from Crypto.Util.Padding import unpad
+from tqdm.rich import tqdm
+
+M3U8_KEY_RE = re.compile(r"((?:#EXT-X-KEY)(?:.*),?URI=\")(?P<url>.*)\",IV=0x(?P<iv>.*)")
+M3U8_MAP_RE = re.compile(r"((?:#EXT-X-MAP)(?:.*),?URI=\")(?P<url>.*)\"(.*)")
+M3U8_SEGMENT_RE = re.compile(r"(?:#EXTINF):.*\n(.*)")
+
+
+def download_hls(m3u8_url, filename, session, threads=5):
+    from .nndownload import FormatNotAvailableException
+
+    with session.get(m3u8_url) as m3u8_request:
+        m3u8_request.raise_for_status()
+        m3u8 = m3u8_request.text
+    key_match = M3U8_KEY_RE.search(m3u8)
+    init_match = M3U8_MAP_RE.search(m3u8)
+    segments = M3U8_SEGMENT_RE.findall(m3u8)
+    if not key_match:
+        raise FormatNotAvailableException("Could not retrieve key file from manifest")
+    if not init_match:
+        raise FormatNotAvailableException("Could not retrieve init file from manifest")
+    if not segments:
+        raise FormatNotAvailableException("Could not retrieve segments from manifest")
+
+    m3u8_type = "video" if "/video/" in m3u8 else "audio"
```

Comment:
Not sure about this... is it keying off of a URL? We should just pass in `stream_type` along with the the M3U8 tuple:

```
    for stream, stream_filename, stream_type in m3u8_streams:
        download_hls(stream, stream_filename, stream_type, session=session, threads=threads)
```